src/librustc/mir/mod.rs

```diff
@@ -212,26 +210,52 @@ impl<'tcx> Body<'tcx> {
 
     #[inline]
     pub fn basic_blocks_mut(&mut self) -> &mut IndexVec<BasicBlock, BasicBlockData<'tcx>> {
-        self.cache.invalidate();
+        self.predecessors_cache = None;
+//        self.cache.invalidate();
         &mut self.basic_blocks
     }
 
     #[inline]
     pub fn basic_blocks_and_local_decls_mut(
         &mut self,
     ) -> (&mut IndexVec<BasicBlock, BasicBlockData<'tcx>>, &mut LocalDecls<'tcx>) {
-        self.cache.invalidate();
+        self.predecessors_cache = None;
+//        self.cache.invalidate();
         (&mut self.basic_blocks, &mut self.local_decls)
     }
 
     #[inline]
-    pub fn predecessors(&self) -> MappedReadGuard<'_, IndexVec<BasicBlock, Vec<BasicBlock>>> {
-        self.cache.predecessors(self)
+    pub fn unwrap_predecessors(&self) -> &IndexVec<BasicBlock, Vec<BasicBlock>> {
+        assert!(self.predecessors_cache.is_some());
+        self.predecessors_cache.as_ref().unwrap()
     }
 
     #[inline]
-    pub fn predecessors_for(&self, bb: BasicBlock) -> MappedReadGuard<'_, Vec<BasicBlock>> {
-        MappedReadGuard::map(self.predecessors(), |p| &p[bb])
+    pub fn predecessors(&mut self) -> &IndexVec<BasicBlock, Vec<BasicBlock>> {
```

Comment:
add a doc comment explaining that this will compute the cache if necessary

Reply:
Does this comment suffice, or do you think we need something more detailed?